### ISSUE TITLE
--enable-alquimia --disable-crunchtope

### ISSUE THREAD

**Issue by ecoon** [amanzi/amanzi#285]
It appears that Alquimia requires building Crunchtope, but superbuild is not aware of that?  Or at least, running bootstrap with: `--enable alquimia --disable-crunchtope` happily builds Alquimia and doesn't build Crunchtope, but then inserts:

```
set(CRUNCHTOPE_DIR ${AMANZI_TPLS_INSTALL_DIR} CACHE PATH "CRUNCHTOPE installation prefix")
```

into amanzi-tpl-config.cmake anyway, and fails to configure Amanzi because it can't find the Crunchtop libraries.

**Comment by ecoon** [comment]
So I still don't think SuperBuild should put a CRUNCHTOPE_DIR in the cmake file, but maybe the issue is actually in Amanzi's CMakeLists.txt.  Specifically, this might get fixed by checking ENABLE_CRUNCHTOPE in AmanziTPL.cmake?

**Comment by lipnikov** [comment]
The logic is that Alquimia needs either Crunch or Pflotran. It should work with one package of both enabled. It should fall otherwise. I think I tested all options ... may be missed something

**Comment by ecoon** [comment]
It generates false positives.  If you "-DENABLE_CRUNCHTOPE=False", it should not call FindPackage(CRUNCHTOPE).

Now, it may be the case that, even if it doesn't find Crunchtope, it warns and doesn't error, and Amanzi still builds correctly.  But, for instance, what if it found an old crufty Crunchtope that wouldn't link?

It's an easy fix -- this will be closed in a minute.

**Comment by ecoon** [comment]
NOTES from pull request:

K: I stopped here. The new script cannot read the existing TPLs installation. I think that Amanzi's config should not specify ENABLE_CRUNCH or ENABLE_PFLOTRAN. It should use FindModule instead as in the original design.

E: 6 is a bit trickier. Currently those lines are set in Amanzi's TPLs config.cmake even if the TPLs were configured with ENABLE_CRUNCHTOPE=False.

Just like every other package that we depend upon, there should be an ENABLE_* set. I agree this needs a bit more work to make sure that SuperBuild sets those flags in the cmake.

Yes, this will break existing TPL installations. I'll take this out of the pull request for now, and reopen the ticket, then do a new pull request that adds this check and fixes SuperBuild to put in the required flags.

K: Probably SuperBuild/CMakeLists.txt should use the same logic in XSDK and non-XSDK branches, i.e. set *_DIR only when the package is enabled

**Comment by ecoon** [comment]
Right.  That is what I intend -- SuperBuild should set ENABLE_CRUNCHTOPE and/or ENABLE_PFLOTRAN as appropriate, and, when these are true, also set CRUNCHTOPE_DIR and/or PFLOTRAN_DIR.  Then, Amanzi should check ENABLE and FindPackage if ENABLE=True.  So user config scripts should not have to change, but this change will break existing TPL installations because of the addition of the variables ENABLE_CRUNCHTOPE and ENABLE_PFLOTRAN to the amanzi_tpls_config.cmake file.  

Note that people can manually add these variables to that file to avoid re-installing TPLs.  E.g:

```
echo "set(ENABLE_PFLOTRAN 1 CACHE BOOL \"Enable PFLOTRAN chemistry engine\")" >> ${AMANZI_TPLS_DIR}/share/cmake/amanzi-tpl-config.cmake
echo "set(ENABLE_CRUNCHTOPE 1 CACHE BOOL \"Enable CRUNCHTOPE chemistry engine\")" >> ${AMANZI_TPLS_DIR}/share/cmake/amanzi-tpl-config.cmake
```

assuming they were installed.  If they weren't installed, there will be no need to change TPLs.

**Comment by lipnikov** [comment]
I am a bit lost.  amanzi-tpl-config.cmake does not need new variables. Even if you specify CRUNCHTOPE_DIR there and have actually no this package, it should be safe, since we use FindPackage(CrunchTope). For some reasons you enforce configuration to fail if FindPackage has failed to find a library. Maybe the recent CMake has changed some default policies.

**Comment by lipnikov** [comment]
What are your cmake version and bootstrap options? I'll try to reproduce this behavior on ubuntu.

**Comment by ecoon** [comment]
My initial bug report was incorrect, it was failing for a different reason (related to #286). 

But I still think that, like every other optional package (e.g. MOAB, MSTK, HYPRE, etc) there should be an ENABLE_<PACKAGE> option, and CRUNCHTOPE_DIR should not be set if it was explicitly not asked for. 

For instance, consider if a developer wanted to test that some chemistry code did a reasonable thing when Crunchtope was not built.  Instead of rebuilding their TPLs without Crunchtope, they ought to be able to just rebuild Amanzi with ENABLE_CRUNCHTOPE=False and still use the same TPLs and get a build of Amanzi without CRUNCHTOPE (currently this would FindPackage successfully and build and Amanzi with Crunchtope).  I have done this with meshes, to make sure that all tests still pass if you build without STK or without MOAB.  We used to have tests that, when built without STK, would still try to test using STK and would throw and fail.

Effectively this is just standardizing how we interact with optional packages.  In my opinion, all optional packages should be controlled by ENABLE_<PACKAGENAME> in amanzi-tpls-config.cmake.  Same as every other optional package we use.

**Comment by lipnikov** [comment]
Your scenario has a flaw. Alquimia should be reconfigured when you are dropping or adding a chemistry package. I have no objections against ENABLE_CRUNCH as long as I it is the optimal parameter and I need to specify only ENABLE_ALQUIMIA.

**Comment by ecoon** [comment]
Closing this because it is fixed in Spack installer, which has a richer way (alquimia's spack package has a variant with/without Crunchtope) and I think that is likely the way we'll end up going...
